File: 057 Insert Intervals.cpp
#include <iostream>
#include <vector>
#include <algorithm>
/*
Insert Interval
https://oj.leetcode.com/problems/insert-interval/
*/
#define DEBUG 0
#define NOT_FOUND -2

struct Interval {
    int start;
    int end;
};
Interval make_pair(int x,int y)
{
    Interval r;
    r.start = x;
    r.end = y;
    return r;
}
typedef std::vector<Interval> List;

void dump(List& input)
{
    for(List::iterator i = input.begin();i!=input.end();i++)
    {
        std::cout<<"("<<i->start<<","<<i->end<<") ";
    }
    printf("\n");
}
List insert(List& input, Interval tag)
{
    // right most interval that includes tag.start (include boundary)
    auto lo = std::lower_bound(input.begin(), input.end(), tag, [](const Interval& lt, const Interval& rt){
        return lt.end < rt.start;
    });

    // right first interval that excludes tag.end
    auto up = std::upper_bound(input.begin(), input.end(), tag, [](const Interval& lt, const Interval& rt){
        return lt.end < rt.start;
    });
    if (lo == up) 
        input.insert(lo, tag);
    else 
    {
        --up;
        tag.start = std::min(tag.start, lo->start);
        tag.end = std::max(tag.end, up->end);
        input.erase(lo, up+1);
        input.insert(lo, tag);
    }
    return input;
}
/*
[1,2],[3,5],[6,7],[8,10],[12,16] : [4,9] 
=> [1,2],[3,10],[12,16]
*/
void test02(void){
    List input;
    input.push_back(make_pair(1,2));
    input.push_back(make_pair(3,5));
    input.push_back(make_pair(6,7));
    input.push_back(make_pair(8,10));
    input.push_back(make_pair(12,16));
    Interval tag = make_pair(5,8);
    
    dump(input);
    insert(input, tag);
    dump(input);
}
int main(int argc, char** argv)
{
    test02();
    return 0;
}

File: 057 Insert Intervals.py
'''
57. Insert Interval
https://leetcode.com/problems/insert-interval/

Given a set of non-overlapping intervals, insert a new interval into the intervals (merge if necessary).

You may assume that the intervals were initially sorted according to their start times.

Example 1:

Input: intervals = [[1,3],[6,9]], newInterval = [2,5]
Output: [[1,5],[6,9]]
Example 2:

Input: intervals = [[1,2],[3,5],[6,7],[8,10],[12,16]], newInterval = [4,8]
Output: [[1,2],[3,10],[12,16]]

Explanation: Because the new interval [4,8] overlaps with [3,5],[6,7],[8,10].
'''
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        res = []
        
        # left and right boundary for overlapped interval
        left, right = newInterval[0], newInterval[1] 
        
        for it in intervals:
            if it[1] < newInterval[0]:
                # 1. Non-overlapping, it is in left branch
                res.append(it)
            elif it[0] > newInterval[1]:
                # 2. Non-overlapping, it is in right branch
                # Need to append the overlapped interval
                if left is not None:
                    res.append([left, right])
                    left = None # indicates overlapped interval is inserted
                res.append(it)
            else:
                # 3. Overlapping now. Update boundaries
                left = min(left, it[0])
                right = max(right, it[1])
        
        if left is not None:
            res.append([left, right])
            
        return res
